lib/posix/pthread_mutex.c

```diff
@@ -9,16 +9,218 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
+
+/*
+ *  Default mutex attrs.
+ */
+static const pthread_mutexattr_t def_attr = {
+	.type = PTHREAD_MUTEX_DEFAULT,
+	.protocol = PTHREAD_PRIO_NONE,
+};
+
+static int acquire_mutex(pthread_mutex_t *m, int timeout)
+{
+	int rc = 0, type, count;
+	pthread_t owner;
+	int key = irq_lock();
+
+	owner = m->owner;
+	type = m->type;
+	count = m->lock_count;
+	irq_unlock(key);
+
+	if (owner == pthread_self()) {
+		if (type == PTHREAD_MUTEX_ERRORCHECK) {
+			return EDEADLK;
+		}
+	}
+
+	if (type == PTHREAD_MUTEX_RECURSIVE) {
+
+		if ((owner == pthread_self()) || (count == 0)) {
+			if (count >= MUTEX_MAX_REC_LOCK) {
+				return EAGAIN;
+			}
+			rc = -k_sem_take(m->sem, timeout);
+			count++;
+		} else {
+			rc = EPERM;
+		}
+	} else {
+		rc = k_sem_take(m->sem, timeout);
+		count = 1;
+	}
+
+	if (rc == 0) {
+		key = irq_lock();
+		m->owner = pthread_self();
+		m->lock_count = count;
+		irq_unlock(key);
+	}
+
+	return rc;
+}
+
+/**
+ * @brief Lock POSIX mutex with non-blocking call.
+ *
+ * See IEEE 1003.1
+ */
 int pthread_mutex_trylock(pthread_mutex_t *m)
 {
-	int key = irq_lock(), ret = EBUSY;
+	return acquire_mutex(m, K_NO_WAIT);
+}
+
+/**
+ * @brief Intialize POSIX mutex.
+ *
+ * See IEEE 1003.1
+ */
+int pthread_mutex_init(pthread_mutex_t *m,
+				     const pthread_mutexattr_t *attr)
+{
+	const pthread_mutexattr_t *mattr;
+
+	if (k_mem_pool_alloc(&posix_mutex_sem, &m->block,
```

Comment:
why use k_mem_pool instead of sys_mem_pool? I would vastly prefer that you use sys_mem_pool for mutex allocations. for one thing, sys_mem_pool is user mode compatible and k_mem_pool never will be.
Also please name this something else, the _sem suffix suggests that this object is a semaphore and it's not.

Reply:
Sure. I have used SYS_MEM_POOL and passed similar arguments as k_mem_pool_alloc but its failing. Is there any calculation I need to consider for MAX , MIN blocks here ? @andrewboie